[fastify + other updates]

robbdempsey opened the pull request:
this builds on the previous work in PR #2  and is also related to Issue #3 

* [x] Use Fastify (https://github.com/nearform/acmeair-nodejs/pull/2)
* [x] Use Promises / async/await where possible - The current code makes extensive use of callbacks.
* [x] Use Pino for logging - The current code uses log4j
* [x] Update dependencies
* [x] /src has been linted standard
* [ ] Identify and Improve performance bottlenecks


This PR does not have the changes for 
* [x] Cloudant
* [ ] Cassandra

cc: @jasnell

jackdclark committed:
initial move of routes to fastify

jackdclark committed:
stream reply in hystrix

jackdclark committed:
finish initial move to fastify, remove express module

jackdclark committed:
fastify request schema validation

Commit:
work-in-progress app moved to fastify-cli/fastify-plugin architecture

robbdempsey committed:
[dojo] update to 1.13.0

robbdempsey committed:
[loader-data] removing old reference

robbdempsey committed:
[settings]

moving all settings to the same file

robbdempsey committed:
[mongo] new module for interacting with mongodb

robbdempsey committed:
[loader] new version

robbdempsey committed:
[plugins] moving config and env to their module

robbdempsey committed:
[routes] splits the routes out

robbdempsey committed:
[services] splits out work to be done

robbdempsey committed:
[config] updated to work with api changes

robbdempsey committed:
[config] bump version of dojo

robbdempsey committed:
[loader] update for api and bump dojo

robbdempsey committed:
[common] update login + logout

robbdempsey committed:
[acmeair-common] fix misspelling

robbdempsey committed:
[customer-profile] update for api

robbdempsey committed:
[flights] updates for api

robbdempsey committed:
[check-in] updates for api

robbdempsey committed:
[app]

changes based on breaking everything out into various routes and services

robbdempsey committed:
[package] tidy up

robbdempsey committed:
[clean-up] remove previous work

robbdempsey committed:
[routes] moved to src/routes

robbdempsey committed:
[loader] moved to src/loader

robbdempsey committed:
[gitignore] update for more recent configs

robbdempsey committed:
[happy-lint]

robbdempsey committed:
[node] sets engine to 8.11.3

robbdempsey committed:
[determine-env]

move db settings to env and enable dotenv in fastify-env
still need to determine if the VCAP items are required

robbdempsey committed:
[settings] moved db specifics to env

robbdempsey committed:
[db] module to let us use multiple db types

robbdempsey committed:
[refactor]

move mongo specifics behind new db module
prep for cloudant'

robbdempsey committed:
[config] switch names to match db

robbdempsey committed:
[flights] switch names to match db

robbdempsey committed:
[services] switch names to match db

robbdempsey committed:
[configure-db] move to config for connection details

robbdempsey committed:
[couch] add details to fastify.config

robbdempsey committed:
[config] switch up on the context

robbdempsey committed:
[loader] using db.names

robbdempsey committed:
[start] dbtype is set in env now

robbdempsey committed:
[db + couch] bring couch up to speed with mongo.

robbdempsey committed:
[happy-lint]

mrister commented on authservice_app.js:
```diff
@@ -41,23 +42,52 @@ logger.info("db type=="+dbtype);
-var app        = express(); 				
-var morgan         = require('morgan');
+var schema = {
+	createToken: {
```
minor: indentation seems really big for this obj.

mrister commented on authservice_app.js:
```diff
@@ -82,71 +112,70 @@ function initDB(){
-	app.listen(port);
+	fastify.listen(port, function (err) {
+    if (err) {
+      logger.error("Error starting server " + err);
```
minor: can the logger maybe take in err as the second argument?

temsa replied:
minor: indent looks strange in this function

mrister commented on netflix/hystrix/index.js:
```diff
@@ -9,51 +10,58 @@ exports.setHystrixMetricsStreamHandlerFactory = function(handler){
-	                } else  {
+			reply.code(503).send({error: 'Can not get instance'});
+		} else if (error) { 
+			console.log("Get instance hit error:"+error);
```
maybe add space after `:` so the error message does not get strictly concatenated to the initial message

robbdempsey replied:
I haven't really focused on this code path.  If it's necessary to move forward I'll bring in your ideas.

mrister commented on package.json:
```diff
@@ -1,26 +1,42 @@
+  "version": "0.0.5",
   "private": true,
+  "engines": {
+    "node": ">=8.11.3"
```
minor: there are a lot of `var`s for such high node version :-)

mrister commented on public/config.html:
```diff
@@ -24,305 +24,287 @@
+          
+        },
+        error: (err) => {
+          console.log(err)
```
minor: console.error maybe?

mrister commented on src/db/couch-db.js:
```diff
@@ -0,0 +1,186 @@
+
+  return new Promise((resolve, reject) => {
+    db.bulk({docs: documents}, (err, response) => {
+      if (err) {
```
minor: could be a bit shorter
```js
db.bulk({docs: documents}, (err, response) => err ? reject(err): resolve(response))
```

mrister commented on src/db/index.js:
```diff
@@ -0,0 +1,111 @@
+  const {collectionName, query} = context
+  const dbService = determineDbService(dbType)
+
+  const results = await dbService.deleteOne(dbClient, collectionName, query)
```
minor, return here?

mrister reviewed (approved):
Noted some minor stuff and in some cases indentation seems too big, but overall LGTM!

temsa commented on dataaccess/mongo/index.js:
```diff
@@ -23,14 +23,14 @@
     var module = {};
 
-	var mongodb = require('mongodb');
+//	var mongodb = require('mongodb');
```
if this is not useful anymore, maybe we can drop the code  ?

temsa commented on public/js/acmeair-common.js:
```diff
@@ -38,77 +38,63 @@ function hideWaitDialog() {
-		dojo.byId("loggedinwelcome").innerHTML = 'Welcome Back ' + loggedinuser;
+	} else {
+		const data = session.split(';')
+		dojo.byId("loggedinwelcome").innerHTML = 'Welcome Back ' + data[2];
```
may be a check on data (size notably) + something like :
```
const loggedInUser = data[2]
dojo.byId("loggedinwelcome").innerHTML = `Welcome Back ${loggedInUser}`;
```
would be great for readability...

Does ES6/7 requirement of #3 extends to the frontend @jasnell ?

jasnell replied:
Incrementally, yes, I'd like to get the front and back end updated. The priority right now, however, is getting the backend updated since that's the part that Node.js' benchmarks actually look :-)

temsa commented on public/index.html:
```diff
@@ -24,7 +24,8 @@
   
   <script type="text/javascript" src="js/acmeair-common.js"></script>
-  <script src="http://ajax.googleapis.com/ajax/libs/dojo/1.7.1/dojo/dojo.js" djConfig="parseOnLoad: true"></script>
+  <script src="//ajax.googleapis.com/ajax/libs/dojo/1.13.0/dojo/dojo.js" djConfig="parseOnLoad: true"></script>
```
I know it is not part of #3 scope, but using Dojo on the frontend in 2018 does not feel trendy at all (it is more 2008 than 2018), should this be changed somehow @jasnell before putting the project forward ?

jasnell replied:
See previous answer :-D https://github.com/nearform/acmeair-nodejs/pull/4#discussion_r197123850 ... replacing dojo is definitely something that should be done, but not in this iteration.

temsa commented on public/js/acmeair-common.js:
```diff
@@ -38,77 +38,63 @@ function hideWaitDialog() {
-			dojo.cookie("loggedinuser", userString, {expires: 5});
+
+			hideLoginWaitDialog();
+			const sessionData = `${response.sessionId};${response.customerId};${response.email}`
```
Why using `;` as a separator ? Maybe using a JSON would be easier (and avoid the `data[2]` in a previous comment). I believe a cookie should be able to store all this data without issue

robbdempsey replied:
I was trying to keep the size of the cookie down.  JSON adds overhead to the size with all the extra characters.

temsa commented on public/js/acmeair-common.js:
```diff
@@ -38,77 +38,63 @@ function hideWaitDialog() {
 	showLoginWaitDialog();
 	
-	var userString = document.getElementById('userId').value;
 	dojo.xhrPost({
```
if the async/await es6/7 of #3  extends to the frontend, maybe this code should typically be impacted ?

temsa commented on public/loader.html:
```diff
@@ -24,176 +24,138 @@
     function loadDB() {
     	showWaitDialog();
-    	dojo.xhrGet({
+    	dojo.xhrPost({
```
async/await es6/7 ?

temsa commented on src/db/couch-db.js:
```diff
@@ -0,0 +1,186 @@
+'use strict'
+
+const dbDestroy = (dbClient, collectionName) => {
```
Maybe this could use Bluebird `promisifyAll` in order to be able to use async/await ?

robbdempsey replied:
I dont see the need for this one to be async.  Am I missing something from a vanilla promise perspective?

temsa commented on authservice_app.js:
```diff
@@ -82,71 +112,70 @@ function initDB(){
+	if (serverStarted) return;
 	serverStarted = true;
-	app.listen(port);
+	fastify.listen(port, function (err) {
```
I believe it could be done in an async/await try/catch manner to match more #3 requirements

temsa reviewed (changes requested):
Some minor concerns (and some questions) in comments but the real issue is running the server as described in the `README.md` does not seems to work and does not give me any error, contrary to `master`:
```console
$ robb-fastify> npm  install
npm WARN ajv-keywords@2.1.1 requires a peer of ajv@^5.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN mongodb-core@1.2.31 requires a peer of kerberos@~0.0 but none is installed. You must install peer dependencies yourself.

audited 1171 packages in 2.852s
found 0 vulnerabilities

 $ robb-fastify> node app.js
$ robb-fastify> git co master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
$ master> npm  install
npm WARN deprecated node-uuid@1.3.3: Use uuid module instead
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
npm WARN mongodb-core@1.2.31 requires a peer of kerberos@~0.0 but none is installed. You must install peer dependencies yourself.

added 108 packages from 75 contributors, removed 384 packages, updated 8 packages, moved 4 packages and audited 224 packages in 8.435s
found 32 vulnerabilities (7 low, 19 moderate, 6 high)
  run `npm audit fix` to fix them, or `npm audit` for details
$ master> node app.js
[2018-06-21 10:47:28.316] [INFO] app - host:port==localhost:9080
[2018-06-21 10:47:28.318] [INFO] app - db type==mongo
[2018-06-21 10:47:28.338] [INFO] routes - Use dataaccess:../dataaccess/mongo/index.js
[2018-06-21 10:47:28.427] [INFO] app - Error connecting to database - exiting process: MongoError: connect ECONNREFUSED 127.0.0.1:27017
[2018-06-21 10:47:28.427] [INFO] app - Initialized database connections
[2018-06-21 10:47:28.432] [INFO] app - Express server listening on port 9080

```

robbdempsey committed:
[old-authservice]

removing old code paths.
if we need to spin out the auth to a seperate service its still doable
with route/auth and route/service.  just needs a server config

robbdempsey committed:
[old-authservice] tidy up

robbdempsey committed:
[Docker]

bring Dockerfile up to more current configuration

robbdempsey committed:
[docker-compose] add configuration

has settings for running acmeair and mongo

robbdempsey committed:
[README]

update README with current settings and usage
update Docker README for current usage

robbdempsey committed:
[old-mongo] removes old code

robbdempsey committed:
[old-cloudant] removes old code

robbdempsey commented:
@temsa I completely agree on the frontend pieces (everything under /public), but I left a lot of the frontend code alone since it does have an older library in place.  @jasnell has some ideas on replacing it, but first pass was to get the backend working with a more modern version.

I've updated the README usage, as well as, the Docker README.

can you try the README steps to get up and running again?

robbdempsey committed:
[tidy-up]

robbdempsey committed:
[standard] remove unneccessary ignore path

temsa reviewed (approved):
LGTM now 🚀

robbdempsey committed:
[package] removing unnecessary dependency

robbdempsey committed:
[ignore] /data

robbdempsey committed:
[logging] add connection details for mongo

robbdempsey committed:
[docker] udpate usage for mongo
